src/ber/mod.rs

```diff
@@ -0,0 +1,619 @@
+//! universal, primitive type with a tag number of 4. So lets decode a
+//! `ModifyDNRequest` (decoding is slightly easier because you're given the
+//! lengths and need not calculate them yourself).
+//!
+//! Full packet:
+//! ```notrust
+//! 30 49 02 01 02 6c 44 04 1d 63 6e 3d 73 6f
+//! 6d 65 75 73 65 72 2c 64 63 3d 65 78 61 6d
+//! 70 6c 65 2c 64 63 3d 63 6f 6d 04 0c 75 69
+//! 64 3d 74 65 73 74 2e 75 73 72 01 01 00 80
+//! 12 64 63 3d 64 6f 65 73 6e 74 2c 64 63 3d
+//! 65 78 69 73 74
+//! ```
+//!
+//! All operations in LDAP are done through the `LDAPMessage` type which has
+//! some additional fields to the ones in `ModifyDNRequest` above, see section
+//! 4.1.1 for the full definition of `LDAPMessage`. So we need to decode that as
+//! well. Looking at the first byte we see `30`, the expected type tag of
+//! `SEQUENCE`, a universal, constructed type with a type tag value of `10`
+//! (hex). Now onto the length: `49`, which has a bit pattern of: 0|1001001.
+//! This means we're dealing with a short definite length and the content is 73
+//! bytes long. Decoding the next portion is the `messageID` field of
+//! `LDAPMessage` which is an integer, length 1, with a value of 2 (`02 01 02`).
+//!
+//! The protocol operation is the next thing up -- our `ModifyDNRequest` of
+//! which we expect to find an application-specific, constructed type with a tag
+//! number of 12, which we do! `6c` has a bit pattern of `01|1|01100` which
+//! confirms we're parsing the right type. The length is again in short form,
+//! `44` which means that our `ModifyDNRequest` content is 68 bytes in total.
+//! `entry` is the first field, which is of type `LDAPDN` (which is reall just
+//! an `OCTET STRING`, remember), so we have an `OCTET STRING` tag, a length of
+//! 29, and the content bytes, which in textual form are:
+//! `cn=someuser,dc=example,dc=com`. `newrdn` is another `OCTET STRING` in
+//! disguise, with the bytes defining it: `04 0c 75 69 64 3d 74 65 73 74 2e 75
+//! 73 72`, easy enough: `uid=test.usr`. `deletoldrdn` is a bool, which has a
+//! tag number of 1, and only has a length of one, whose value is `00`. LDAP
+//! defines boolean `TRUE` to be `FF`, so this is `FALSE`. `newSuperior` is an
+//! optional context-specific `LDAPDN`. So we check to see if we're at the end
+//! of the message yet -- nope! still have more bytes to process, then we check
+//! the tag: `80` has a bit pattern of `10|0|00000` which tells us its a
+//! context-specific, universal type with a tag number of 0, that's our type!
+//! The length and content are decoded as usual. We're still not done however,
+//! we finished decoding `ModifyDNRequest`, but we still have the `controls`
+//! field of `LDAPMessage`, which is also an optional type. Check one: have we
+//! reached the end of the packet? As a matter of fact, yes, we have! Therefore
+//! `controls` is `None` and we've finished decoding the entire packet!
```

Comment:
This section is very helpful. For someone just entering the crate, thoughts on whether we could bisect this into two or more distinct portions?

One possible position to split could be:
> So we check to see if we're at the end of the message yet -- nope!

Reply:
Yeah after I wrote the whole thing I wanted to try and split it up slightly more, but wasn't sure where the best place would be. Perhaps I'll even try rewording a couple things slightly so it flows better in paragraph form.